controllers/core/pod_controller.go

```diff
@@ -189,3 +197,32 @@ func (r *PodReconciler) SetupWithManager(ctx context.Context, manager ctrl.Manag
 		MaxConcurrentReconciles: MaxPodConcurrentReconciles,
 	}).UsingConditions(r.Condition).Complete(r)
 }
+
+// updateResourceName updates resource name according to pod event and which IP allocation mode is enabled
+func (r *PodReconciler) updateResourceName(isDeletionEvent bool, pod *v1.Pod) string {
+	resourceName := config.ResourceNameIPAddress
+	// Pod deletion must use the handler that assigned the IP resource regardless which IP allocation mode is active
+	if isDeletionEvent {
+		resourceID, present := pod.Annotations[resourceName]
+		if !present {
+			r.Log.Error(fmt.Errorf("resource ID was not found in annotation"),
+				"resource", resourceName, "is delete event", isDeletionEvent)
+			return resourceName
+		}
+		// Check prefix provider to see if it's a prefix deconstructed IP
+		providerMap := r.ResourceManager.GetResourceProviders()
+		prefixProvider := providerMap[config.ResourceNameIPAddressFromPrefix]
```

Comment:
The pod controller is async reconciler. I am not sure if there is safe to read from the map without locking. But adding locks in reconciling need cautions. Can you check you other change in Manager and see if can read from there and protected by a lock?

Reply:
Actually I can't do it in resource manage since it doesn't have a lock there. It's always locked by the caller

Reply:
Pushed a new commit to add the lock for resource map in resource manager, and a new API to get resource provider by name.

Reply:
Removing the newly added lock here as discussed offline.